### ISSUE TITLE
Performance issue in CI

### ISSUE THREAD

**Issue by n-rodriguez** [fxn/zeitwerk#12]
Hi there!

I don't know if it's the right place for this issue, but Rails bug tracker is very busy so I open it here :)

My issue : I've tested my Rails app with Rails 6.0.0.beta2 and tried the new Zeitgest loader, and noticed that my test suite duration is twice longer than before (in Gitlab CI env)  :/

The context : 

1. I use [grosser/parallel_tests](https://github.com/grosser/parallel_tests) to run the test suite with 6 cpus, and usually it takes 6/7mn to run.

2. Last week I introduced Zeitwerk in a bunch of gems I use ([simple_navigation_bootstrap](https://github.com/jbox-web/simple_navigation_bootstrap), [active_settings](https://github.com/jbox-web/active_settings)) and noticed that the tests were a bit longer but didn't dig into it.

3. Finally, I've set up Zeitwerk in 9 gems that I use in this app, and noticed that the test suite was really longer : it took 12/13mn to run.

4. Now I use Zeitwerk in the Rails app and the test suite take 30mn to run :cry: 

By reverting the Rails loader to the classic mode it takes 13mn to run.

I could remove Zeitwerk in the gems I use and run some tests and I'm pretty sure it will solve the issue.
But I don't think it's the right direction to go.

Maybe I should eager load everything in CI env?

What do you think?

Thank you!

**Comment by n-rodriguez** [comment]
I put `Zeitwerk::Loader.eager_load_all` in an initializer with Rails classic loader, so all gems using Zeitwerk were eager loaded and the test suite took 6mn :tada:

**Comment by n-rodriguez** [comment]
Okay! I think I got it :

* switch Rails loader to Zeitwerk mode
* add `Zeitwerk::Loader.eager_load_all` in `spec/rails_helper.rb`

This will eager load all gems that use Zeitwerk and the Rails application itself but only when doing `bundle exec rspec` or when using parallel_test.

Now the test suite duration is back to 6/7mn :tada:

**Comment by fxn** [comment]
Eager loading does basically the same as autoloading, only it does it upfront, instead of on demand. In theory, total wall clock would be the same, but your app refutes that so let's see why is that.

Since you have several gems plus the app (hooray!), let's see if the several tracers listening to the `:class` event are adding up. Let's try something easy. Could you please `bundle show zeitwerk` and comment out [this line](https://github.com/fxn/zeitwerk/blob/74b727b7c814d3bd2f65cc84087b54091ab45e2b/lib/zeitwerk/loader.rb#L271) in the method that eager loads?

With that edition done, run the suite as you have it now, eager loading all.

Let's see if it is still as fast or the slow down is back.

**Comment by n-rodriguez** [comment]
> With that edition done, run the suite as you have it now, eager loading all.
> 
> Let's see if it is still as fast or the slow down is back.

The slow down is back. The test suite took 16mn to run.

**Comment by fxn** [comment]
Cool. So one hypothesis could be that having 11 trace points slows down the whole thing, possibly because there are many modules and classes being created at boot or whatever.

I am going to try to reproduce this in a synthetic test case with for example 100 dummy gems using Zetiwerk and one app with a good number of dependencies (that will define a good number of classes).

If I can reproduce and confirm the hypothesis, then there is a plan B which is to have one single global trace point. It would be fixable, but I want to understand it first.

I'll write back!

**Comment by n-rodriguez** [comment]
> I am going to try to reproduce this in a synthetic test case with for example 100 dummy gems using Zeitwerk and one app with a good number of dependencies (that will define a good number of classes).

:+1: 

> I'll write back!

Thank you!

**Comment by fxn** [comment]
I have created a `Gemfile` with some 150 gems, and this test file:

```ruby
lazy_subdirs = {}

tracers = []
100.times do
  tracer = TracePoint.new(:class) do |tp|
    lazy_subdirs.delete(tp.self.name)
  end
  tracer.enable
  tracers << tracer
end

require "bundler/setup"
Bundler.require
```

This test creates or reopens classes and modules 20K times, resulting in 2M calls to the trace point block. That results in an almost 5x slow down. The test runs in about 48s, while commenting out the trace point setup results in 10s.

The time goes into the computation of the hash code for the class or module name, since `Hash#delete` needs it to perform the deletion. If we replace the code in the block with with something cheaper like `lazy_subdirs.empty?(tp.self.name)`, the 100 trace points have a negligible cost.

This justifies creating one single global trace point instead of one per loader, since I want projects to be able to depend on 100 gems using Zeitwerk.

I'll work on a patch that you can test.

**Comment by n-rodriguez** [comment]
> I'll work on a patch that you can test.

Thanks!

**Comment by fxn** [comment]
There is a proof of concept (WIP) in the `global-tp` branch:

```ruby
gem 'zeitwerk', git: 'git@github.com:fxn/zeitwerk.git', branch: 'global-tp'
```

Would you like to give it a try?

**Comment by n-rodriguez** [comment]
> Would you like to give it a try?

Should I remove eager loading in `spec/rails_helper.rb`?

**Comment by n-rodriguez** [comment]
With your change the test suite took about 8mn to run. 

It's a bit more than with Rails 5 but we're far from the 18mn of the beginning.

Also everything was eager loaded. I'm gonna run the test suite without eager loading.

**Comment by fxn** [comment]
Yes please, do not force eager loading. With this patch we reduce the number of trace points from 11 to 1.

Coincidentally, I just force pushed an update, please run a bundle update, current SHA1 is `f8a6df5`.

**Comment by n-rodriguez** [comment]
With your latest change and without eager loading, the test suite took about 6mn to run. We're back in normal test duration :)

Files are loaded faster 
before : `files took 12.58 seconds to load`
after : `files took 6.87 seconds to load` 

but Zeitwerk seems to add a small overhead of 1mn comparatively with Rails loader 
before : `Took 294 seconds (4:54) # Rails 5 + eager load gems`
after : `Took 371 seconds (6:11) # Rails 6 + patch, no eager load`

**Comment by fxn** [comment]
Excellent. I'll polish and release a new version.

Does Rals 5 + eager load gems means that your gems using Zeitwerk eager load in their main file regardless of the Rails app? Do they do the same in the Rails 6 case, only the app is not eager loaded?

**Comment by n-rodriguez** [comment]
> Does Rals 5 + eager load gems mean that your gems using Zeitwerk eager load in their main file regardless of the Rails app?

Nope. I do this in Rails in an initializer : 

```ruby
# Eager load gems that use Zeitwerk.
# This should be reviewed when upgrading to Rails 6
# See: https://github.com/fxn/zeitwerk/issues/12
Zeitwerk::Loader.eager_load_all
```

**Comment by n-rodriguez** [comment]
In Rails 6 I've replaced the initializer with 

```ruby
autoloader = Rails.autoloaders.main
sti_leaves = %w[campings customers email_campaigns prospects sale_agreements sent_booklets visit_bills users]

sti_leaves.each do |leaf|
  autoloader.preload Rails.root.join('app', 'concepts', 'charts', 'operation', "#{leaf}.rb").to_s
end
```

**Comment by fxn** [comment]
Perfect.

It could be the case that the classic autoloader is a tad lighter, because the classic autoloader is reactive. It does nothing on boot, and when `const_missing` is triggered then it does its thing.

Zeitwerk mode does as little as possible, but it is not nothing. For example, it has to scan the first level of files to set autoloads, and it might need a trace point if if finds explicit namespaces. That trace point is called by Ruby whenever a class or module is created or reopened. That is not very expensive, and we carefully are as lazy as possible, but it is more than nothing.

So Zeitwerk is efficient _assuming a correct solution to the problem_. Classic is a tad lighter, but does not autoload following Ruby semantics. You know, you can make anything fast f it can be wrong :).

**Comment by n-rodriguez** [comment]
> For example, it has to scan the first level of files to set autoloads, and it might need a trace point if it finds explicit namespaces. That trace point is called by Ruby whenever a class or module is created or reopened. That is not very expensive, and we carefully are as lazy as possible, but it is more than nothing.

That's what I thought. 

But don't get me wrong, it was not a criticism but more an information about what to expect when using Zeitwerk ;) 

IMHO Zeitwerk is (one of) the missing pieces in the Ruby ecosystem and I'm very glad to use it :+1:

**Comment by fxn** [comment]
I published `1.3.2` with that refactor, could you please confirm it behaves as expected?

**Comment by n-rodriguez** [comment]
> I published `1.3.2` with that refactor, could you please confirm it behaves as expected?

It works! Thank you!

**Comment by fxn** [comment]
Awesome, thanks for opening this issue and for your help in understanding this.
